File: 2015/02/cpp/solution.cpp

#include <algorithm>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

int part_1(vector<string> input) {
  int total = 0;

  for (string line : input) {
    int i = line.find('x');
    int x = stoi(line.substr(0, i));
    string rem = line.substr(i + 1);
    i = rem.find('x');
    int y = stoi(rem.substr(0, i));
    rem = rem.substr(i + 1);
    int z = stoi(rem);

    int dims[] = {x, y, z};
    int n = sizeof(dims) / sizeof(dims[0]);

    sort(dims, dims + n);

    total += dims[0] * dims[1];

    total += 2 * x * y + 2 * x * z + 2 * y * z;
  }

  return total;
}

int part_2(vector<string> input) {
  int total = 0;

  for (string line : input) {
    int i = line.find('x');
    int x = stoi(line.substr(0, i));
    string rem = line.substr(i + 1);
    i = rem.find('x');
    int y = stoi(rem.substr(0, i));
    rem = rem.substr(i + 1);
    int z = stoi(rem);

    int dims[] = {x, y, z};
    int n = sizeof(dims) / sizeof(dims[0]);
    sort(dims, dims + n);

    total += dims[0] + dims[0] + dims[1] + dims[1];
    total += dims[0] * dims[1] * dims[2];
  }

  return total;
}

int main() {
  vector<string> input;
  ifstream file("../input.txt");
  string line;
  while (getline(file, line)) {
    input.push_back(line);
  }
  file.close();

  cout << "Part 1: " << part_1(input) << endl;
  cout << "Part 2: " << part_2(input) << endl;
}

File: 2015/02/python/solution.py
with open("../input.txt") as file:
    lines = file.readlines()


def part_1(data: list[str]) -> int:
    total = 0
    for line in data:
        tokens = line.split("x")
        values = [int(i) for i in tokens]
        values.sort()
        total += values[0] * values[1]
        total += (
            2 * values[0] * values[1]
            + 2 * values[1] * values[2]
            + 2 * values[2] * values[0]
        )
    return total


def part_2(data: list[str]) -> int:
    total = 0
    for line in data:
        tokens = line.split("x")
        values = [int(i) for i in tokens]
        values.sort()
        total += values[0] + values[0] + values[1] + values[1]
        total += values[0] * values[1] * values[2]
    return total


if __name__ == "__main__":
    print(f"Part 1: {part_1(lines)}")
    print(f"Part 2: {part_2(lines)}")
